apps/web/scripts/kernel/handlers/solanaHandler.ts

```diff
@@ -37,47 +38,126 @@ export const handleSignMessage: MessengerCallback = async (
 	payload,
 	channel,
 ) => {
-	const privateKey = await triggerActionToGetPrivateKey();
-
-	if (!privateKey) {
-		return;
-	}
-
-	const message = decode(payload.message);
-	const signature = signMessage(message, privateKey);
-
-	channel.postMessage({
-		from: 'walless@kernel',
-		requestId: payload.requestId,
-		signature: encode(signature),
-	});
-
-	return signature;
+	const { requestId = '' } = payload;
+	const popup = await handleOpenPopup(PopupType.SIGNATURE_POPUP, requestId);
+	requestMap[requestId] = {
+		channel,
+		payload,
+		popup,
+		resolve: false,
+	};
 };
 
-export const handleSignTransaction: MessengerCallback = async (
+export const handleResolveRequestSignature: MessengerCallback = async (
 	payload,
 	channel,
 ) => {
-	const privateKey = await triggerActionToGetPrivateKey();
+	const requestSource = requestMap[payload.requestId as string];
+	const responsePayload: ResponsePayload = {
+		from: 'walless@kernel',
+		requestId: payload.requestId,
+	};
 
-	if (!privateKey) {
-		return;
-	}
+	const handleRejectRequest = (message: string) => {
+		requestSource.channel.postMessage({
+			from: 'walless@kernel',
+			requestId: requestSource.payload.requestId,
+			message,
+		});
+
+		channel.postMessage({
+			from: 'walless@kernel',
+			requestId: payload.requestId,
+			responseCode: ResponseCode.ERROR,
+			message: 'Request rejected',
+		});
```

Comment:
Continue to previous comment, I think it might be refactored like:
```typescript
messaging.repsonse({ 
          to: requestId, 
          responseCode: ResponseCode.ERROR, 
          payload 
})
```
- `to` and `responseCode` is required, and `from` will be attached from var which is initalized with messaging
- this `response` method will look at message pool, then we can pass over multiple channels without messy and conflict.
@FME849 @fluctlight-kayaba I mean we need to redifine this